impeller/renderer/backend/vulkan/binding_helpers_vk.cc

```diff
@@ -113,156 +118,111 @@ static bool BindBuffers(const Bindings& bindings,
+                   write_workspace, write_offset) ||
+      !BindImages(command.fragment_bindings, allocator, encoder, descriptor_set,
+                  image_workspace, image_offset, write_workspace,
+                  write_offset)) {
+    return fml::Status(fml::StatusCode::kUnknown,
+                       "Failed to bind texture or buffer.");
   }
 
-  // Step 2: Update the descriptors for all image and buffer descriptors used
-  // in the render pass.
-  std::vector<vk::DescriptorImageInfo> images;
-  std::vector<vk::DescriptorBufferInfo> buffers;
-  std::vector<vk::WriteDescriptorSet> writes;
-  images.reserve(samplers_count + subpass_count);
-  buffers.reserve(buffer_count);
-  writes.reserve(samplers_count + buffer_count + subpass_count);
-
-  auto& allocator = *context.GetResourceAllocator();
-  auto desc_index = 0u;
-  for (const auto& command : commands) {
-    auto desc_set = command.pipeline->GetDescriptor()
-                        .GetVertexDescriptor()
-                        ->GetDescriptorSetLayouts();
-    if (!BindBuffers(command.vertex_bindings, allocator, encoder,
-                     descriptor_sets[desc_index], desc_set, buffers, writes) ||
-        !BindBuffers(command.fragment_bindings, allocator, encoder,
-                     descriptor_sets[desc_index], desc_set, buffers, writes) ||
-        !BindImages(command.fragment_bindings, allocator, encoder,
-                    descriptor_sets[desc_index], images, writes)) {
-      return fml::Status(fml::StatusCode::kUnknown,
-                         "Failed to bind texture or buffer.");
-    }
+  if (pipeline_descriptor.UsesSubpassInput()) {
+    vk::DescriptorImageInfo image_info;
+    image_info.imageLayout = vk::ImageLayout::eGeneral;
+    image_info.sampler = VK_NULL_HANDLE;
+    image_info.imageView = input_attachment.GetImageView();
+    image_workspace[image_offset++] = image_info;
 
-    if (command.pipeline->GetDescriptor().UsesSubpassInput()) {
-      vk::DescriptorImageInfo image_info;
-      image_info.imageLayout = vk::ImageLayout::eGeneral;
-      image_info.sampler = VK_NULL_HANDLE;
-      image_info.imageView = input_attachment.GetImageView();
-      images.push_back(image_info);
-
-      vk::WriteDescriptorSet write_set;
-      write_set.dstSet = descriptor_sets[desc_index];
-      write_set.dstBinding = kMagicSubpassInputBinding;
-      write_set.descriptorCount = 1u;
-      write_set.descriptorType = vk::DescriptorType::eInputAttachment;
-      write_set.pImageInfo = &images.back();
-
-      writes.push_back(write_set);
-    }
-    desc_index += 1;
+    vk::WriteDescriptorSet write_set;
+    write_set.dstSet = descriptor_set;
+    write_set.dstBinding = kMagicSubpassInputBinding;
+    write_set.descriptorCount = 1u;
+    write_set.descriptorType = vk::DescriptorType::eInputAttachment;
+    write_set.pImageInfo = &image_workspace[image_offset - 1];
+
+    write_workspace[write_offset++] = write_set;
```

Comment:
Other than the writes past the end. This reads so much less scarier. Rather than depending on the correct `reserve` call to ensure that the `push_back` doesn't cause a reallocation and invalid pointer.